### PR TITLE
[PF-580] Fix Harness abort lifecycle settlement

### PR DESCRIPTION
**Pull request by mbenhamd**
## Summary

- Settle legacy Harness sendMessage when abort wins before signal acceptance, without allowing the deferred admission task to start an orphan run.
- Emit a single aborted agent_end for legacy subscribed streams that ignore abort, with active-subscription guards for mid-chunk races.
- Make Harness v1 active-turn abort waiters settle non-cooperative turns and map agent_end reasons for aborted/error completions across message, signal, reminder, and queue paths.
- Add v1 regression coverage for an agent that ignores abort.

## Validation

- pnpm prettier:changed
- pnpm --filter @mastra/core check
- pnpm --filter mastracode check
- pnpm --filter ./packages/core test:unit src/harness/v1/session.scheduler.test.ts src/harness/v1/session.cancel.test.ts src/harness/v1/session.tool-context.test.ts src/harness/v1/session.suspend.test.ts src/harness/v1/workspace-restore.test.ts src/harness/v1/contracts.test.ts src/harness/v1/token-usage.test.ts src/harness/v1/session.abort.test.ts src/harness/om-failure-abort.test.ts src/harness/signal-messages.test.ts --reporter=dot
  - 10 files, 195 tests passed. token-usage stderr is expected deliberate failure injection.
- pnpm --filter @mastra/core build:lib
- pnpm --filter ./mastracode test src/headless-integration.test.ts src/harness/runtime.test.ts src/harness/events.test.ts src/tui/handlers/__tests__/prompts.test.ts src/agents/__tests__/model.test.ts src/tools/__tests__/request-sandbox-access.test.ts --reporter=dot
  - 6 files, 116 tests passed.
- pnpm --filter mastracode build:lib
- git diff --check

## Review Evidence

- Claude review returned no BLOCKER findings. One valid legacy duplicate-event REVIEW finding was addressed by adding active-subscription guards after stream chunk processing and in the stream error path.
- Codex review pass 1 found blockers around v1 non-message turn agent_end coverage and legacy follow-up draining; both were fixed.
- Codex review pass 2 found a legacy pre-dispatch orphan-run blocker and v1 queued-turn lifecycle issues; these were fixed with abort-generation admission guards, queued agent_end tracking, and direct non-cooperative abort coverage.
- CodeRabbit CLI was attempted twice, including coderabbit review --agent --base main; it connected and entered summarizing but did not return findings, so this PR should rely on the GitHub CodeRabbit app review.

Linear: PF-580

ELI5
This PR makes the Harness reliably stop in-progress runs when asked to abort: runs are cleaned up deterministically, duplicate terminal events are avoided, and orphaned or continued work (including from agents that ignore aborts) is prevented.

Changes Overview

Core Harness Abort Lifecycle (packages/core/src/harness/harness.ts)
- Introduces generation-scoped abort tracking (abortGeneration, abortWaiters) to detect and gate concurrent aborts and to race admission against aborts.
- Makes sendSignal/sendMessage admission generation-aware; sendMessage now races admission against abort waiters and ensures deterministic cleanup in finally blocks.
- Adds active-subscription guards and repeated active-checks around chunk processing and error paths to avoid mid-chunk races and duplicate/late processing after unsubscribe/abort.
- Refactors direct-stream and subscribed-stream handling with generation-aware waiters, run-finalization markers, and helpers to avoid multiple agent_end emissions and orphan runs.
- Reworks abort() to increment abortGeneration, resolve abort waiters, best-effort abort active subscriptions/controllers, clear pending tool approvals, and drive a single, well-guarded agent_end with reason 'aborted' for legacy/direct flows.
- Reworks tool approval/resume plumbing to use run-aware AbortControllers, add run-level in-flight protections, prevent resuming when another run is active, and clear pending approval state after handling.

Session Lifecycle & Agent-End Semantics (packages/core/src/harness/v1/session.ts)
- Stores per-turn AbortController and ties waiter rejection to that controller via _scheduleActiveTurnAbort so only active-turn waiters are rejected.
- Normalizes agent_end.reason via _agentEndReasonForFullOutput and centralizes emission; adds flags (agentEndEmitted/streamAgentEndEmitted) to guarantee a single terminal event across message, signal, reminder, and queue paths.
- Updates queued-run completion to propagate onAgentEnd callbacks and avoid duplicate agent_end emission during queued-finalization.

Tests and Coverage
- Adds tests for agents that ignore aborts (AbortIgnoringMockAgent) and asserts session.message() rejects with HarnessSessionCancelledError while emitting exactly one agent_end with reason 'aborted' (v1/session.abort.test.ts).
- Expands harness tool suspension/resumption tests to cover direct-resume abort cleanup, many race conditions, follow-up draining, retargeting, and pending-approval edge cases (harness-tool-suspension.test.ts).
- Tightens and refactors multiple session tests: message abort semantics, queued-run finalization behavior, mode-switch abort wait semantics, and tool-context slot abort signal expectations.
- Test runs reported: unit suites across updated files (195 tests passed; token-usage stderr noted as deliberate), integration suites: 116 tests passed. Prettier, package checks/builds for `@mastra/core` and mastracode, and git diff --check were executed.

Review & Validation
- Addressed automated/code review findings:
  - Claude review: no blockers; added active-subscription guards for duplicate-event case.
  - Codex reviews (pass 1 & 2): addressed blockers for non-message turn agent_end coverage, legacy follow-up draining, pre-dispatch orphan-run, and queued-turn lifecycle issues (applied fixes: abort-generation admission guards, queued agent_end tracking, direct non-cooperative abort coverage).
  - CodeRabbit CLI attempts were made; final review via GitHub CodeRabbit app.
- Linked Linear ticket: PF-580
- Changeset updated to document improved Harness v1 session cancellation reliability (.changeset/pf-747-harness-cancellation.md).

Impact / Notes
- No public API or exported signature changes.
- Significant internal changes and added tests; review effort classified as high due to complex abort/resume/race handling across multiple run types.

[Review Change Stack](https://app.coderabbit.ai/change-stack/mbenhamd/mastra/pull/191?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/core/src/harness/harness.ts [hunk 1]
```diff
@@ -2539,16 +2597,33 @@ export class Harness<TState = {}> {
...
+      } catch {}
+      if (this.agentThreadSubscription === activeSubscription) {
+        this.agentThreadSubscription = null;
+        this.agentThreadSubscriptionKey = null;
+      }
+      this.currentRunId = null;
+      this.currentTraceId = null;
+      this.abortRequested = false;
+      this.emit({ type: 'agent_end', reason: 'aborted' });
+    }
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/harness.ts:2916]: Agent-authored reply with user authorization: addressed in cd24af7a68. The abort path now routes subscribed aborts through handleSubscribedStreamError(), which performs standard cleanup and drains follow-ups, and direct resume abort finalization drains queued follow-ups before releasing the abort marker. Added regression coverage in harness-tool-suspension.test.ts for direct abort follow-up draining and multiple queued follow-ups. Resolving as fixed.

File: packages/core/src/harness/v1/session.ts [hunk 2]
```diff
@@ -1152,6 +1153,23 @@ export class Session {
     }
   }
 
+  private _scheduleActiveTurnAbort(controller: AbortController): void {
+    const reason = (controller.signal as { reason?: unknown }).reason;
+    const message = typeof reason === 'string' && reason.length > 0 ? reason : 'session_aborted';
+    const err = reason instanceof Error ? reason : new HarnessSessionCancelledError(this.id, message);
+    setTimeout(() => {
+      if (this._currentTurnAbortController !== controller || !controller.signal.aborted) return;
+      this._rejectActiveTurnWaiters(err);
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/session.ts]: Agent-authored reply with user authorization: addressed in cd24af7a68. _scheduleActiveTurnAbort now snapshots waiters for the aborted controller and rejects that snapshot even if turn ownership changes before the timeout fires. Added coverage for completed caller-abort cleanup and abort waiter settlement. Resolving as fixed.

File: packages/core/src/harness/v1/session.ts [hunk 3]
```diff
@@ -8217,7 +8287,19 @@ export class Session {
...
+      const full = await Promise.race([completion, activeTurnWaiter.promise]);
+      agentEndEmitted = true;
+      return full;
+    } catch (err) {
+      if (agentStarted && !agentEndEmitted) {
+        this._emitTurnEvent({
+          type: 'agent_end',
+          reason: turnAbortController.signal.aborted ? 'aborted' : 'error',
+          runId: identity.runId,
+          queuedItemId: item.id,
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/session.ts:8322]: Agent-authored reply with user authorization: addressed in cd24af7a68. _runQueuedTurn now tracks fallbackRunId from the resolved signalIdentity and uses it for the fallback agent_end emission, so remapped dispatch IDs stay correlated. Resolving as fixed.

File: packages/core/src/harness/v1/session.ts [hunk 4]
```diff
@@ -8347,7 +8430,7 @@ export class Session {
         activeTurnWaiter,
       );
     } else {
-      await this._finalizeQueuedRunCompletion(item, full, modeId, activeTurnWaiter);
+      await this._finalizeQueuedRunCompletion(item, full, modeId, activeTurnWaiter, opts);
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/session.ts:8456]: Agent-authored reply with user authorization: addressed in cd24af7a68. _awaitQueuedRunCompletion now forwards opts through the non-suspended _finalizeCompletedQueuedTurn path, and _finalizeCompletedQueuedTurn passes onAgentEnd into _finalizeQueuedRunCompletion. Added queue regression coverage asserting only one queued agent_end. Resolving as fixed.

File: packages/core/src/harness/harness.ts [hunk 5]
```diff
@@ -2539,16 +2597,33 @@ export class Harness<TState = {}> {
...
+        activeSubscription?.unsubscribe();
+      } catch {}
+      if (this.agentThreadSubscription === activeSubscription) {
+        this.agentThreadSubscription = null;
+        this.agentThreadSubscriptionKey = null;
+      }
+      this.currentRunId = null;
+      this.currentTraceId = null;
+      this.abortRequested = false;
+      this.emit({ type: 'agent_end', reason: 'aborted' });
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/harness.ts:2915]: Agent-authored reply with user authorization: addressed in cd24af7a68. Direct resume aborts are now tracked separately from subscribed streams with abortFinalizedRunIds/directStreamRunIds, abortRequested is not cleared in abort(), and processStream/handleToolResume observe the abort marker before cleaning up. Added direct-resume abort tests for pre-chunk, pending resumeStream, approval wait, partial messages, and stale shutdown admission. Resolving as fixed.

File: packages/core/src/harness/harness.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/harness.ts:2916]: Agent-authored reply with user authorization: addressed in cd24af7a68. abort() now distinguishes subscribed-thread aborts from direct resume streams: subscribed aborts use handleSubscribedStreamError() cleanup/drain behavior, while direct resume aborts use the direct stream marker/runtime abort path. Added regression coverage for null activeRunId subscribed aborts and direct resume follow-up/runtime release. Resolving as fixed.

File: packages/core/src/harness/v1/session.ts [hunk 6]
```diff
@@ -8336,7 +8442,7 @@ export class Session {
 
     if (full.finishReason !== 'suspended') {
       await this._raceActiveTurnWaiter(this._markQueuedTurnCompleted(item.id, full), activeTurnWaiter);
-      await this._finalizeCompletedQueuedTurn(item, full, modeId, activeTurnWaiter);
+      await this._finalizeCompletedQueuedTurn(item, full, modeId, activeTurnWaiter, opts);
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/session.ts:8445]: Agent-authored reply with user authorization: addressed in 34bd8ef0d3. The queued-turn fallback `agent_end` now skips `QueuePostRunFinalizationPendingError`, because that sentinel means the run already completed and the queued head is parked for retry; the retry finalization path remains the single owner of the terminal `agent_end`. Extended `session.queue.test.ts` so the post-run finalization retry case asserts zero `agent_end` before retry and exactly one after finalization. Validation run: `pnpm --filter @mastra/core test src/harness/v1/session.queue.test.ts` passed (51 tests, no type errors), `pnpm --filter @mastra/core check` passed, `pnpm prettier:changed` passed, `git diff --check` passed. Two local Codex review passes found no issues; local CodeRabbit CLI was attempted but stopped after several minutes without findings because it hung before producing output. Resolving as fixed.

**Comment by mbenhamd** [comment]
@coderabbitai review

Please refresh the review/status for the current head `34bd8ef0d3d312859e6b1b35319c2abcfad34273`. The latest CodeRabbit thread was confirmed addressed, all review threads are resolved, but the separate CodeRabbit status context is still pending.
